File: webapp/src/types/ws.ts
export enum WSMsgType {
  // Server => Client messages [1, 1000]
  WSMsgSrvError = 1,
  WSMsgSrvProctreeDump,
  WSMsgSrvSchedhookCap,
  WSMsgSrvVMEDump,

  // Client => Server messages [1001, +inf]
  WSMsgClientReqProctreeDump = 1001,
  WSMSgClientReqSchedhookCapStart,
  WSMsgClientReqSchedhookCapEnd,
  WSMsgClientReqVMEDump,
}

export interface WSMessage {
  type: WSMsgType;
  payload: any;
}

export type ListenerFn = (payload: any) => void;

export function isWSMessage(obj: any): obj is WSMessage {
  return (
    obj !== null &&
    typeof obj === "object" &&
    typeof obj.type === "number" &&
    "payload" in obj
  );
}

export enum WSStatus {
  Disconnected,
  Connecting,
  Connected,
  Error,
}

File: bridge/model/ws.go
package model

import "encoding/json"

type WSMsgType uint32

// Server => Client messages [1, 1000]
const (
	WSMsgSrvError WSMsgType = iota + 1
	WSMsgSrvProctreeDump
	WSMsgSrvSchedhookCap
	WSMsgSrvVMEDump
)

// Client => Server messages [1001, +inf]
const (
	WSMsgClientReqProctreeDump WSMsgType = iota + 1001
	WSMSgClientReqSchedhookCapStart
	WSMsgClientReqSchedhookCapEnd
	WSMsgClientReqVMEDump
)

func NewWSMsg(msgType WSMsgType, payload any) (WSMessage, error) {
	msg := WSMessage{
		Type:    msgType,
		Payload: nil,
	}

	if payload == nil {
		return msg, nil
	}

	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return WSMessage{}, err
	}
	msg.Payload = payloadBytes

	return msg, nil
}

type WSMessage struct {
	Type    WSMsgType       `json:"type"`
	Payload json.RawMessage `json:"payload"`
}

const WSChannelBufferSize = 1000

type WSContext struct {
	WriteCh chan<- WSMessage
}
